### PR TITLE
stdenv: add alderlake support

### PR DESCRIPTION
**Pull request by CHN-beta**
###### Description of changes

Add alderlake support, so that user could build system with

```
nixpkgs.hostPlatform = { system = "x86_64-linux"; gcc = { arch = "alderlake"; tune = "alderlake"; }; };
```

While testing on my computer, some packages do broken if build with `alderlake` been set. But it might still meaningful to users who only want some of packages in the system being built with `-march=alderlake`. Meanwhile this commit should not cause more fail if `alderlake` is not set.

I have not tested all packages as my machine is not so powerful. But at least gcc and llvm stuff works fine with `alderlake` been set. I only found 3 broken packages after built over 1000 derivations in the past days with globally set `nixpkgs.hostPlatform` to the value above. The broken packages are `python-aiohttp` (a test failed to run), `openexr` (a test failed to run) and `mono` (failed to build).

I am new to NixOS, and I am sorry if I made any mistake in contributing to nixpkgs.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: lib/systems/architectures.nix [hunk 1]
```diff
@@ -55,6 +56,7 @@ rec {
     cascadelake    = [ "skylake-avx512" ] ++ inferiors.cannonlake;
     cooperlake     = [ "cascadelake"    ] ++ inferiors.cascadelake;
     tigerlake      = [ "icelake-server" ] ++ inferiors.icelake-server;
+    alderlake      = [ "syklake"        ] ++ inferiors.tigerlake;
```
Comment (lucasew) [inline lib/systems/architectures.nix:59]: skylake?

**Review by lucasew — CHANGES_REQUESTED** [review]

**Comment by lucasew** [comment]
It looks good. I also don't know very much about this part of nixpkgs tho.

File: lib/systems/architectures.nix (hunk 1 above)
Comment (CHN-beta) [reply lib/systems/architectures.nix:59]: Sorry for the typo. It should be `[ "skylake" ] ++ inferiors.skylake`.

However I noticed that according to [gcc document](https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html), there exist instructions supported by all old cpus (from westmere to tigerlake) but not supported by alderlake, such as `CX16`. I am not an expert in CPU nor nixpkgs, should alderlake be a superior CPU of skylake in such conditions, or should alderlake be a superior CPU of none of old series?

Comment (lucasew) [reply lib/systems/architectures.nix]: Typos often create bugs but that's the cool thing about code review: More people looking at the code before it's shipped, so these typos are more likely to be caught. Less pressure for the author to make things perfect too.

**Comment by SuperSandro2000** [comment]
Can you also add rocketlake and graniterapids ?

**Comment by CHN-beta** [comment]
> Can you also add rocketlake and graniterapids ?

Sorry, I am not expert in CPU architectures and does not have machines in rocketlake or graniterapids to run tests. Might the work is better be done by other experienced people.

**Comment by SuperSandro2000** [comment]
Most of those machines where added without real hardware and based on the linked resources like https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html

**Comment by lucasew** [comment]
> Most of those machines where added without real hardware and based on the linked resources like https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html

Isn't there any structured file like a JSON that can be used to generate these options automatically?

**Comment by CHN-beta** [comment]
> > Most of those machines where added without real hardware and based on the linked resources like https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html
> 
> Isn't there any structured file like a JSON that can be used to generate these options automatically?

Might better to just write cpu architectures and their supported instructions into nix, and let nix automatically decide which is inferior over another.

I am new to NixOS, and it is a good chance for me to exercise nix language and to contribute to community. I would try to do the job this weekend if nobody does it before. But I am not expert in such instructions and really does not know what the `CX16` or `FMA` things do. I would try my best to not make anything break down.

**Comment by CHN-beta** [comment]
I am trying to add support for `rocketlake` and `graniterapids` but find something really puzzled me.

In the current version of nixpkgs, `znver1` was listed as an inferior over `skylake`. However there exist instructions supported by `skylake` but not supported by `znver1` (such as `SAHF`). I wonder if there is something important I do not realize, maybe `SAHF` was renamed into some other name in AMD CPUs, or it was never used in practice (like something in [x32](https://en.wikipedia.org/wiki/X32_ABI) stuff).

Another confusion is [here](https://github.com/NixOS/nixpkgs/blob/42637bd82223ba0e8c21e8ce5ec6b905b2967593/lib/systems/architectures.nix#L57), the expression before `++` seems no effect as `cannonlake` has already been an inferior over `skylake-avx512`. This may be also a typo mistake?

By the way, I find the `inferiors` attribute set was only used by nix-daemon and only used once ([here](https://github.com/NixOS/nixpkgs/blob/42637bd82223ba0e8c21e8ce5ec6b905b2967593/nixos/modules/services/misc/nix-daemon.nix#L832)).

So I think the corresponding code would be better to be refactored. It might better to just write cpu architectures and their supported instructions into nix, as well as some information about instuctions (such as some of them may never be used in practice), and let nix automatically decide which cpu type is inferior over another. But this may take a lot of time to do and really need an expertise to review the code. Let me make a pr when I have more spare time.

**Comment by SuperSandro2000** [comment]
Those could be certainly mistakes...

And the code you pointed at is also the only usecase for that function.

**Pull request merged**